File: backend/app/schemas/deal.py
import uuid
from datetime import date, datetime

from pydantic import BaseModel, computed_field


class DealCreate(BaseModel):
    farmer_id: uuid.UUID
    buyer_id: uuid.UUID
    product_id: uuid.UUID
    quantity: float
    unit: str = "kg"
    buy_rate: float
    sell_rate: float
    transport_cost: float = 0
    labour_cost: float = 0
    other_cost: float = 0
    transporter_id: uuid.UUID | None = None
    deal_date: date | None = None
    delivery_date: date | None = None
    payment_due_date: date | None = None
    quality_grade: str | None = None
    notes: str | None = None


class DealUpdate(BaseModel):
    farmer_id: uuid.UUID | None = None
    buyer_id: uuid.UUID | None = None
    product_id: uuid.UUID | None = None
    quantity: float | None = None
    unit: str | None = None
    buy_rate: float | None = None
    sell_rate: float | None = None
    transport_cost: float | None = None
    labour_cost: float | None = None
    other_cost: float | None = None
    deal_date: date | None = None
    payment_due_date: date | None = None
    quality_grade: str | None = None
    transporter_id: uuid.UUID | None = None
    status: str | None = None
    farmer_payment_status: str | None = None
    buyer_payment_status: str | None = None
    farmer_paid_amount: float | None = None
    buyer_received_amount: float | None = None
    spoilage_qty: float | None = None
    spoilage_reason: str | None = None
    delivery_date: date | None = None
    has_dispute: bool | None = None
    dispute_notes: str | None = None
    notes: str | None = None


class DealResponse(BaseModel):
    id: uuid.UUID
    farmer_id: uuid.UUID
    buyer_id: uuid.UUID
    product_id: uuid.UUID
    quantity: float
    unit: str
    buy_rate: float
    sell_rate: float
    transport_cost: float
    labour_cost: float
    other_cost: float
    status: str
    farmer_payment_status: str
    buyer_payment_status: str
    farmer_paid_amount: float
    buyer_received_amount: float
    spoilage_qty: float
    spoilage_reason: str | None
    transporter_id: uuid.UUID | None
    deal_date: date
    delivery_date: date | None
    payment_due_date: date | None
    quality_grade: str | None
    has_dispute: bool
    dispute_notes: str | None
    notes: str | None
    created_at: datetime

    # Related names (populated in router)
    farmer_name: str | None = None
    buyer_name: str | None = None
    product_name: str | None = None

    @computed_field
    @property
    def buy_total(self) -> float:
        return self.quantity * self.buy_rate

    @computed_field
    @property
    def sell_total(self) -> float:
        return self.quantity * self.sell_rate

    @computed_field
    @property
    def gross_margin(self) -> float:
        return self.sell_total - self.buy_total

    @computed_field
    @property
    def total_cost(self) -> float:
        return self.transport_cost + self.labour_cost + self.other_cost

    @computed_field
    @property
    def net_profit(self) -> float:
        return self.gross_margin - self.total_cost

    model_config = {"from_attributes": True}

File: backend-spring/src/main/java/com/nearkart/entity/Deal.java
package com.nearkart.entity;

import jakarta.persistence.*;
import lombok.*;
import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.OffsetDateTime;
import java.util.UUID;

@Entity
@Table(name = "deals")
@Getter @Setter
@NoArgsConstructor @AllArgsConstructor
@Builder
public class Deal {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    @Column(name = "user_id", nullable = false)
    private UUID userId;

    @Column(name = "farmer_id")
    private UUID farmerId;

    @Column(name = "buyer_id")
    private UUID buyerId;

    @Column(name = "product_id", nullable = false)
    private UUID productId;

    @Column(nullable = false, precision = 10, scale = 2)
    private BigDecimal quantity;

    @Builder.Default
    private String unit = "kg";

    @Column(name = "buy_rate", nullable = false, precision = 10, scale = 2)
    private BigDecimal buyRate;

    @Column(name = "sell_rate", nullable = false, precision = 10, scale = 2)
    private BigDecimal sellRate;

    @Column(name = "transport_cost", precision = 10, scale = 2)
    @Builder.Default
    private BigDecimal transportCost = BigDecimal.ZERO;

    @Column(name = "labour_cost", precision = 10, scale = 2)
    @Builder.Default
    private BigDecimal labourCost = BigDecimal.ZERO;

    @Column(name = "other_cost", precision = 10, scale = 2)
    @Builder.Default
    private BigDecimal otherCost = BigDecimal.ZERO;

    @Builder.Default
    private String status = "pending";

    @Column(name = "farmer_payment_status")
    @Builder.Default
    private String farmerPaymentStatus = "unpaid";

    @Column(name = "buyer_payment_status")
    @Builder.Default
    private String buyerPaymentStatus = "unpaid";

    @Column(name = "farmer_paid_amount", precision = 14, scale = 2)
    @Builder.Default
    private BigDecimal farmerPaidAmount = BigDecimal.ZERO;

    @Column(name = "buyer_received_amount", precision = 14, scale = 2)
    @Builder.Default
    private BigDecimal buyerReceivedAmount = BigDecimal.ZERO;

    @Column(name = "spoilage_qty", precision = 10, scale = 2)
    @Builder.Default
    private BigDecimal spoilageQty = BigDecimal.ZERO;

    @Column(name = "spoilage_reason")
    private String spoilageReason;

    @Column(name = "transporter_id")
    private UUID transporterId;

    @Column(name = "deal_date")
    private LocalDate dealDate;

    @Column(name = "delivery_date")
    private LocalDate deliveryDate;

    @Column(name = "payment_due_date")
    private LocalDate paymentDueDate;

    @Column(name = "quality_grade")
    private String qualityGrade;

    @Column(name = "has_dispute")
    @Builder.Default
    private Boolean hasDispute = false;

    @Column(name = "dispute_notes", columnDefinition = "TEXT")
    private String disputeNotes;

    @Column(columnDefinition = "TEXT")
    private String notes;

    @Column(name = "created_at", nullable = false, updatable = false)
    private OffsetDateTime createdAt;

    @Column(name = "updated_at", nullable = false)
    private OffsetDateTime updatedAt;

    // Computed properties (DB generated columns - read-only)
    @Column(name = "buy_total", insertable = false, updatable = false)
    private BigDecimal buyTotal;

    @Column(name = "sell_total", insertable = false, updatable = false)
    private BigDecimal sellTotal;

    @Column(name = "gross_margin", insertable = false, updatable = false)
    private BigDecimal grossMargin;

    @Column(name = "total_cost", insertable = false, updatable = false)
    private BigDecimal totalCost;

    @Column(name = "net_profit", insertable = false, updatable = false)
    private BigDecimal netProfit;

    // Relationships
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "farmer_id", insertable = false, updatable = false)
    private Farmer farmer;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "buyer_id", insertable = false, updatable = false)
    private Buyer buyer;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "product_id", insertable = false, updatable = false)
    private Product product;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "transporter_id", insertable = false, updatable = false)
    private Transporter transporter;

    @PrePersist
    protected void onCreate() {
        OffsetDateTime now = OffsetDateTime.now();
        createdAt = now;
        updatedAt = now;
        if (dealDate == null) dealDate = LocalDate.now();
    }

    @PreUpdate
    protected void onUpdate() {
        updatedAt = OffsetDateTime.now();
    }

    // Transient computed fields for when DB generated columns aren't available
    @Transient
    public BigDecimal computeBuyTotal() {
        return quantity != null && buyRate != null ? quantity.multiply(buyRate) : BigDecimal.ZERO;
    }

    @Transient
    public BigDecimal computeSellTotal() {
        return quantity != null && sellRate != null ? quantity.multiply(sellRate) : BigDecimal.ZERO;
    }

    @Transient
    public BigDecimal computeGrossMargin() {
        return computeSellTotal().subtract(computeBuyTotal());
    }

    @Transient
    public BigDecimal computeTotalCost() {
        BigDecimal t = transportCost != null ? transportCost : BigDecimal.ZERO;
        BigDecimal l = labourCost != null ? labourCost : BigDecimal.ZERO;
        BigDecimal o = otherCost != null ? otherCost : BigDecimal.ZERO;
        return t.add(l).add(o);
    }

    @Transient
    public BigDecimal computeNetProfit() {
        return computeGrossMargin().subtract(computeTotalCost());
    }
}
